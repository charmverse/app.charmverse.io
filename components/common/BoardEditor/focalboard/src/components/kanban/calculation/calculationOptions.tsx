
import type { IPropertyTemplate, PropertyType } from '../../../blocks/board';
import type {
  CommonCalculationOptionProps } from '../../calculations/options';
import {
  CalculationOptions,
  optionsByType
} from '../../calculations/options';

import type { OptionProps } from './kanbanOption';
import { Option } from './kanbanOption';

type Props = CommonCalculationOptionProps & {
    cardProperties: IPropertyTemplate[];
    onChange: (data: { calculation: string, propertyId: string }) => void;
    property: IPropertyTemplate;
}

// contains mapping of property types which are effectly the same as other property type.
const equivalentPropertyType = new Map<PropertyType, PropertyType>([
  ['createdTime', 'date'],
  ['updatedTime', 'date']
]);

export function getEquivalentPropertyType (propertyType: PropertyType): PropertyType {
  return equivalentPropertyType.get(propertyType) || propertyType;
}

export function KanbanCalculationOptions (props: Props): JSX.Element {
  const options: OptionProps[] = [];

    // Show common options, first,
    // followed by type-specific functions
    optionsByType.get('common')!.forEach((typeOption) => {
      if (typeOption.value !== 'none') {
        options.push({
          ...typeOption,
          cardProperties: props.cardProperties,
          onChange: props.onChange,
          activeValue: props.value,
          activeProperty: props.property
        });
      }
    });

    const seen: Record<string, boolean> = {};
    props.cardProperties.forEach((property) => {
      // skip already processed property types
      if (seen[getEquivalentPropertyType(property.type)]) {
        return;
      }

      (optionsByType.get(property.type) || [])
        .forEach((typeOption) => {
          options.push({
            ...typeOption,
            cardProperties: props.cardProperties,
            onChange: props.onChange,
            activeValue: props.value,
            activeProperty: props.property!
          });
        });

      seen[getEquivalentPropertyType(property.type)] = true;
    });

    return (
      <CalculationOptions
        value={props.value}
        menuOpen={props.menuOpen}
        onClose={props.onClose}
        onChange={props.onChange}
        options={options}
        components={{ Option }}
      />
    );
}
