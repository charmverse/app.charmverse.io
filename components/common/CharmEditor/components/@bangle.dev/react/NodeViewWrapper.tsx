import type { NodeView, NodeViewProps } from '@bangle.dev/core';
import React from 'react';

export type RenderNodeViewsFunction = (
  props: NodeViewProps & { children: React.ReactNode },
) => React.ReactNode;

interface PropsType {
  nodeView: NodeView;
  renderNodeViews: RenderNodeViewsFunction;
  nodeViewUpdateStore: WeakMap<NodeView, () => void>;
}

interface StateType {
  nodeViewProps: NodeViewProps;
}

export class NodeViewWrapper extends React.Component<PropsType, StateType> {
  update: () => void;

  attachToContentDOM: (reactElement: HTMLDivElement) => void;

  constructor (props: PropsType) {
    super(props);
    this.update = () => {
      this.setState(() => ({
        nodeViewProps: props.nodeView.getNodeViewProps()
      }));
    };

    this.attachToContentDOM = (reactElement) => {
      if (!reactElement) {
        return;
      }
      const contentDOM = this.props.nodeView.contentDOM!;
      // Since we do not control how many times this callback is called
      // make sure it is not already mounted.
      if (!reactElement.contains(contentDOM)) {
        // If contentDOM happens to be mounted to someone else
        // remove it from there.
        if (contentDOM.parentNode) {
          contentDOM.parentNode.removeChild(contentDOM);
        }
        reactElement.appendChild(contentDOM);
      }
    };
    // So that we can directly update the nodeView without the mess
    // of prop forwarding.
    // What about updating the wrong nodeView ?
    // It is okay because a nodeView and this ReactComponent will always
    // have a 1:1 mapping. This is guaranteed because you use `nodeView` instance
    // to generate a react key. See the usage of this component in ./ReactEditor.js
    props.nodeViewUpdateStore.set(props.nodeView, this.update);
    this.state = { nodeViewProps: this.props.nodeView.getNodeViewProps() };
  }

  getChildren () {
    if (!this.props.nodeView.contentDOM) {
      return null;
    }

    // if (this.state.nodeViewProps.node.isInline) {
    //   return (
    //     // The bangle-nv-content is needed to keep the content take space
    //     // or else browsers will collapse it, making it hard to type
    //     <span
    //       className="bangle-nv-child-container"
    //       ref={this.attachToContentDOM}
    //     />
    //   );
    // }

    return (
      <div
        className='bangle-nv-child-container'
        ref={this.attachToContentDOM}
      />
    );
  }

  render () {
    const element = this.props.renderNodeViews({
      ...this.state.nodeViewProps,
      children: this.getChildren()
    });
    if (!element) {
      throw new Error(
        `renderNodeView must handle rendering for node of type "${this.state.nodeViewProps.node.type.name}"`
      );
    }
    return element;
  }
}
