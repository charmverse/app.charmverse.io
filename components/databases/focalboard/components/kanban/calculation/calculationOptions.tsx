// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'

import {
    CalculationOptions,
    CommonCalculationOptionProps,
    optionsByType,
} from '../../calculations/options'
import {IPropertyTemplate, PropertyType} from '../../../blocks/board'

import './calculationOption.scss'
import {Option, OptionProps} from './kanbanOption'

type Props = CommonCalculationOptionProps & {
    cardProperties: IPropertyTemplate[]
    onChange: (data: {calculation: string, propertyId: string}) => void
}

// contains mapping of property types which are effectly the same as other property type.
const equivalentPropertyType = new Map<PropertyType, PropertyType>([
    ['createdTime', 'date'],
    ['updatedTime', 'date'],
])

export function getEquivalentPropertyType(propertyType: PropertyType): PropertyType {
    return equivalentPropertyType.get(propertyType) || propertyType
}

export const KanbanCalculationOptions = (props: Props): JSX.Element => {
    const options: OptionProps[] = []

    // Show common options, first,
    // followed by type-specific functions
    optionsByType.get('common')!.forEach((typeOption) => {
        if (typeOption.value !== 'none') {
            options.push({
                ...typeOption,
                cardProperties: props.cardProperties,
                onChange: props.onChange,
                activeValue: props.value,
                activeProperty: props.property!,
            })
        }
    })

    const seen: Record<string, boolean> = {}
    props.cardProperties.forEach((property) => {
        // skip already processed property types
        if (seen[getEquivalentPropertyType(property.type)]) {
            return
        }

        (optionsByType.get(property.type) || []).
            forEach((typeOption) => {
                options.push({
                    ...typeOption,
                    cardProperties: props.cardProperties,
                    onChange: props.onChange,
                    activeValue: props.value,
                    activeProperty: props.property!,
                })
            })

        seen[getEquivalentPropertyType(property.type)] = true
    })

    return (
        <CalculationOptions
            value={props.value}
            menuOpen={props.menuOpen}
            onClose={props.onClose}
            onChange={props.onChange}
            property={props.property}
            options={options}
            components={{Option}}
        />
    )
}
