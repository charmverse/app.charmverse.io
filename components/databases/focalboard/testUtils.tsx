// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'
import {DndProvider} from 'react-dnd'
import {HTML5Backend} from 'react-dnd-html5-backend'
import configureStore, {MockStoreEnhanced} from 'redux-mock-store'
import {Middleware} from 'redux'

import {Block} from './blocks/block'

export const wrapIntl = (children?: React.ReactNode): JSX.Element => <IntlProvider locale='en'>{children}</IntlProvider>
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
    return (
        <DndProvider backend={HTML5Backend}>
            {wrapIntl(children)}
        </DndProvider>
    )
}

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()
        range.getBoundingClientRect = jest.fn()
        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }
        return range
    }
}
export function mockMatchMedia(result: {matches: boolean}): void {
    // We check if system preference is dark or light theme.
    // This is required to provide it's definition since
    // window.matchMedia doesn't exist in Jest.
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation(() => {
            return result

            // return ({
            //     matches: true,
            // })
        }),
    })
}

export function mockStateStore(middleware:Middleware[], state:unknown): MockStoreEnhanced<unknown, unknown> {
    const mockStore = configureStore(middleware)
    return mockStore(state)
}

export type BlocksById<BlockType> = {[key: string]: BlockType}

export function blocksById<BlockType extends Block>(blocks: Array<BlockType>): BlocksById<BlockType> {
    return blocks.reduce((res, block) => {
        res[block.id] = block
        return res
    }, {} as BlocksById<BlockType>)
}
