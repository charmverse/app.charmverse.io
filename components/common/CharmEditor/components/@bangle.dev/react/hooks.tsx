import type { EditorView } from '@bangle.dev/pm';
import { Plugin, PluginKey } from '@bangle.dev/pm';
import { rafSchedule } from '@bangle.dev/utils';
import { useContext, useEffect, useState } from 'react';

import type { BangleEditorStateProps } from 'components/common/CharmEditor/components/@bangle.dev/core/bangle-editor-state';

import { BangleEditorState } from '../core/bangle-editor-state';

import { EditorViewContext } from './ReactEditor';

export function useEditorState(props: BangleEditorStateProps) {
  if (props.plugins && typeof props.plugins !== 'function') {
    throw new Error('plugins error: plugins must be a function');
  }

  const [state] = useState(
    () =>
      // Instantiate the editorState once and keep using that instance
      // on subsequent renders.
      // Passing a callback in useState lazy calls the
      // functions on the first render and never again.
      new BangleEditorState(props)
  );

  return state;
}

export function usePluginState(pluginKey: PluginKey, throttle = false) {
  const view = useEditorViewContext();
  const [state, setState] = useState(pluginKey.getState(view.state));

  useEffect(() => {
    let _setState = setState;
    if (throttle) {
      _setState = rafSchedule(setState);
    }
    const plugin = watcherPlugin(pluginKey, _setState);
    (view as any)._updatePluginWatcher(plugin);
    return () => {
      if (throttle) {
        (_setState as ReturnType<typeof rafSchedule>).cancel();
      }
      (view as any)._updatePluginWatcher(plugin, true);
    };
  }, [view, pluginKey, throttle]);

  return state;
}

export function useEditorViewContext(): EditorView {
  const context = useContext(EditorViewContext);
  if (!context) {
    throw new Error('useEditorViewContext must be used within a EditorViewContext');
  }
  return context;
}

function watcherPlugin(pluginKey: PluginKey, setState: (state: any) => void) {
  return new Plugin({
    key: new PluginKey(`withPluginState_${(pluginKey as any).key}`),
    view() {
      return {
        update(view, prevState) {
          const { state } = view;
          if (prevState === state) {
            return;
          }
          const newPluginState = pluginKey.getState(state);

          if (newPluginState !== pluginKey.getState(prevState)) {
            setState(newPluginState);
          }
        }
      };
    }
  });
}
