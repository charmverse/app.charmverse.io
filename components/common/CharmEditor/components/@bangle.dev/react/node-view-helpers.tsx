import { objectUid } from '@bangle.dev/utils';
import { log } from '@charmverse/core/log';
import { useEffect, useState } from 'react';
import type { RefObject } from 'react';
import { flushSync } from 'react-dom';

import type { NodeView, RenderHandlers } from '../core/node-view';
import { saveRenderHandlers } from '../core/node-view';

export const nodeViewUpdateStore = new WeakMap();

type NodeViewsUpdater = (nodeViewUpdateStore: NodeView[]) => NodeView[];
type UpdateNodeViewsFunction = (updater: NodeViewsUpdater) => void;

export const nodeViewRenderHandlers = (updateNodeViews: UpdateNodeViewsFunction): RenderHandlers => ({
  create: (nodeView, _nodeViewProps) => {
    log.debug('create node', objectUid.get(nodeView)); // , new Error().stack);
    updateNodeViews((nodeViews) => [...nodeViews, nodeView]);
  },
  update: (nodeView, _nodeViewProps) => {
    log.debug('update node', objectUid.get(nodeView));
    const updateCallback = nodeViewUpdateStore.get(nodeView);
    // If updateCallback is undefined (which can happen if react took long to mount),
    // we are still okay, as the latest nodeViewProps will be accessed whenever it mounts.
    if (updateCallback) {
      updateCallback();
    }
  },
  destroy: (nodeView) => {
    // log.debug('destroy', objectUid.get(nodeView));
    updateNodeViews((nodeViews) => nodeViews.filter((n) => n !== nodeView));
  }
});

export function useNodeViews(ref: RefObject<HTMLElement>) {
  const [nodeViews, setNodeViews] = useState<NodeView[]>([]);
  useEffect(() => {
    // save the renderHandlers in the dom to decouple nodeView instantiating code
    // from the editor. Since PM passing view when nodeView is created, the author
    // of the component can get the handler reference from `getRenderHandlers(view)`.
    // Note: this assumes that the pm's dom is the direct child of `editorRenderTarget`.
    let destroyed = false;
    saveRenderHandlers(
      ref.current!,
      nodeViewRenderHandlers((cb) => {
        if (!destroyed) {
          // make sure that updates run sequentially for prosemiror, or we get infinite recursion of node views being created and destroyed.
          // See also: https://github.com/ueberdosis/tiptap/pull/2985
          flushSync(() => {
            // use callback variant of setState to
            // always get freshest nodeViews.
            setNodeViews((_nodeViews) => cb(_nodeViews));
          });
        }
      })
    );
    return () => {
      destroyed = true;
    };
  }, [ref]);

  return nodeViews;
}
