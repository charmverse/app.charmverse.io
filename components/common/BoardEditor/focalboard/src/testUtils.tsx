import type { PageMeta } from '@charmverse/core/pages';
import React from 'react';
import { DndProvider } from 'react-dnd';
import { HTML5Backend } from 'react-dnd-html5-backend';
import { IntlProvider } from 'react-intl';
import type { Middleware } from 'redux';
import type { MockStoreEnhanced } from 'redux-mock-store';
import configureStore from 'redux-mock-store';
import { v4 } from 'uuid';

import { PagesContext } from 'hooks/usePages';
import type { Block } from 'lib/focalboard/block';
import { pageStubToCreate } from 'testing/generatePageStub';
import { AppThemeProvider } from 'theme/AppThemeProvider';

export const wrapIntl = (children?: React.ReactNode): JSX.Element => (
  <AppThemeProvider>
    <IntlProvider locale='en'>{children}</IntlProvider>
  </AppThemeProvider>
);
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
  return <DndProvider backend={HTML5Backend}>{wrapIntl(children)}</DndProvider>;
};

export const wrapPagesProvider = (cardId: string | string[], children: React.ReactNode) => {
  const pages: Record<string, PageMeta> = {};

  const cardIds = Array.isArray(cardId) ? cardId : [cardId];

  cardIds.forEach((_cardId) => {
    pages[_cardId] = pageStubToCreate({ spaceId: v4(), createdBy: v4(), id: _cardId, path: _cardId }) as PageMeta;
  });

  return (
    <PagesContext.Provider
      value={
        {
          pages
        } as any
      }
    >
      {children}
    </PagesContext.Provider>
  );
};

export function mockDOM(): void {
  window.focus = jest.fn();
  document.createRange = () => {
    const range = new Range();
    range.getBoundingClientRect = jest.fn();
    range.getClientRects = () => {
      return {
        item: () => null,
        length: 0,
        [Symbol.iterator]: jest.fn()
      };
    };
    return range;
  };
}
export function mockMatchMedia(result: { matches: boolean }): void {
  // We check if system preference is dark or light theme.
  // This is required to provide it's definition since
  // window.matchMedia doesn't exist in Jest.
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(() => {
      return result;

      // return ({
      //     matches: true,
      // })
    })
  });
}

export function mockStateStore(middleware: Middleware[], state: unknown): MockStoreEnhanced<unknown, unknown> {
  const mockStore = configureStore(middleware);
  return mockStore(state);
}

export type BlocksById<BlockType> = { [key: string]: BlockType };

export function blocksById<BlockType extends Block>(blocks: BlockType[]): BlocksById<BlockType> {
  return blocks.reduce((res, block) => {
    res[block.id] = block;
    return res;
  }, {} as BlocksById<BlockType>);
}
