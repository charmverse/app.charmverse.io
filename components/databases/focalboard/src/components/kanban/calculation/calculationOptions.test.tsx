// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'

import {render} from '@testing-library/react'

import userEvent from '@testing-library/user-event'

import {TestBlockFactory} from '../../../test/testBlockFactory'
import {wrapIntl} from '../../../testUtils'

import {KanbanCalculationOptions} from './calculationOptions'

describe('components/kanban/calculations/KanbanCalculationOptions', () => {
    const board = TestBlockFactory.createBoard()

    test('base case', () => {
        const component = wrapIntl(
            <KanbanCalculationOptions
                value={'count'}
                property={board.fields.cardProperties[1]}
                menuOpen={false}
                onChange={() => {}}
                cardProperties={board.fields.cardProperties}
            />,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('with menu open', () => {
        const component = wrapIntl(
            <KanbanCalculationOptions
                value={'count'}
                property={board.fields.cardProperties[1]}
                menuOpen={true}
                onChange={() => {}}
                cardProperties={board.fields.cardProperties}
            />,
        )

        const {container} = render(component)
        expect(container).toMatchSnapshot()
    })

    test('with submenu open', () => {
        const component = wrapIntl(
            <KanbanCalculationOptions
                value={'count'}
                property={board.fields.cardProperties[1]}
                menuOpen={true}
                onChange={() => {}}
                cardProperties={board.fields.cardProperties}
            />,
        )

        const {container, getByText} = render(component)
        const countUniqueValuesOption = getByText('Count Unique Values')
        expect(countUniqueValuesOption).toBeDefined()
        userEvent.hover(countUniqueValuesOption)
        expect(container).toMatchSnapshot()
    })

    test('duplicate property types', () => {
        const boardWithProps = TestBlockFactory.createBoard()
        boardWithProps.fields.cardProperties.push({
            id: 'number-property-1',
            name: 'A Number Property - 1',
            type: 'number',
            options: [],
        })
        boardWithProps.fields.cardProperties.push({
            id: 'number-property-2',
            name: 'A Number Propert - 2y',
            type: 'number',
            options: [],
        })

        const component = wrapIntl(
            <KanbanCalculationOptions
                value={'count'}
                property={boardWithProps.fields.cardProperties[1]}
                menuOpen={true}
                onChange={() => {}}
                cardProperties={boardWithProps.fields.cardProperties}
            />,
        )

        const {getAllByText} = render(component)
        const sumOptions = getAllByText('Sum')
        expect(sumOptions).toBeDefined()
        expect(sumOptions.length).toBe(1)
    })

    test('effectively date fields', () => {
        // date, created time and updated time are effectively date fields.
        // Only one set of date related menus should show up for all of them.

        const boardWithProps = TestBlockFactory.createBoard()
        boardWithProps.fields.cardProperties.push({
            id: 'date',
            name: 'Date',
            type: 'date',
            options: [],
        })
        boardWithProps.fields.cardProperties.push({
            id: 'created-time',
            name: 'Created Time',
            type: 'createdTime',
            options: [],
        })
        boardWithProps.fields.cardProperties.push({
            id: 'updated-time',
            name: 'Updated Time',
            type: 'updatedTime',
            options: [],
        })

        const component = wrapIntl(
            <KanbanCalculationOptions
                value={'count'}
                property={boardWithProps.fields.cardProperties[1]}
                menuOpen={true}
                onChange={() => {}}
                cardProperties={boardWithProps.fields.cardProperties}
            />,
        )

        const {getAllByText} = render(component)

        const earliestDateMenu = getAllByText('Earliest Date')
        expect(earliestDateMenu).toBeDefined()
        expect(earliestDateMenu.length).toBe(1)

        const latestDateMenu = getAllByText('Latest Date')
        expect(latestDateMenu).toBeDefined()
        expect(latestDateMenu.length).toBe(1)

        const dateRangeMenu = getAllByText('Date Range')
        expect(dateRangeMenu).toBeDefined()
        expect(dateRangeMenu.length).toBe(1)
    })
})
