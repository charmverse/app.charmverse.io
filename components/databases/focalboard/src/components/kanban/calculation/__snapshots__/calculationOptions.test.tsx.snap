// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/kanban/calculations/KanbanCalculationOptions base case 1`] = `
<div>
  <div
    class="CalculationOptions css-2b097c-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-context"
      >
          Select is focused , press Down to open the menu, 
      </span>
    </span>
    <div
      class="CalculationOptions__control CalculationOptions__control--is-focused css-1s59geg-Control"
    >
      <div
        class="CalculationOptions__value-container CalculationOptions__value-container--has-value css-1mxrbau-ValueContainer"
      >
        <div
          class="CalculationOptions__single-value css-1brck82-singleValue"
        >
          Count
        </div>
        <input
          aria-autocomplete="list"
          class="css-wmatm6-dummyInput-DummyInput"
          id="react-select-2-input"
          readonly=""
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="CalculationOptions__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="CalculationOptions__indicator CalculationOptions__clear-indicator css-13eygzs-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class="CalculationOptions__indicator-separator css-43ykx9-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="CalculationOptions__indicator CalculationOptions__dropdown-indicator css-1f9iddu-indicatorContainer"
        >
          <i
            class="CompassIcon icon-chevron-up ChevronUpIcon"
          />
        </div>
      </div>
    </div>
    <input
      name="calculation_options"
      type="hidden"
      value="count"
    />
  </div>
</div>
`;

exports[`components/kanban/calculations/KanbanCalculationOptions with menu open 1`] = `
<div>
  <div
    class="CalculationOptions css-2b097c-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-context"
      >
         7 results available. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <div
      class="CalculationOptions__control CalculationOptions__control--is-focused CalculationOptions__control--menu-is-open css-1s59geg-Control"
    >
      <div
        class="CalculationOptions__value-container CalculationOptions__value-container--has-value css-1mxrbau-ValueContainer"
      >
        <div
          class="CalculationOptions__single-value css-1brck82-singleValue"
        >
          Count
        </div>
        <input
          aria-autocomplete="list"
          class="css-wmatm6-dummyInput-DummyInput"
          id="react-select-3-input"
          readonly=""
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="CalculationOptions__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="CalculationOptions__indicator CalculationOptions__clear-indicator css-13eygzs-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class="CalculationOptions__indicator-separator css-43ykx9-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="CalculationOptions__indicator CalculationOptions__dropdown-indicator css-1f9iddu-indicatorContainer"
        >
          <i
            class="CompassIcon icon-chevron-up ChevronUpIcon"
          />
        </div>
      </div>
    </div>
    <div
      class="CalculationOptions__menu css-1rsmi4x-menu"
    >
      <div
        class="CalculationOptions__menu-list css-g29tl0-MenuList"
      >
        <div
          class="KanbanCalculationOptions_CustomOption active"
        >
          <span>
            Count
             
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Not Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Percent Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Percent Not Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Value
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Unique Values
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
      </div>
    </div>
    <input
      name="calculation_options"
      type="hidden"
      value="count"
    />
  </div>
</div>
`;

exports[`components/kanban/calculations/KanbanCalculationOptions with submenu open 1`] = `
<div>
  <div
    class="CalculationOptions css-2b097c-container"
  >
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      class="css-1f43avz-a11yText-A11yText"
    >
      <span
        id="aria-selection"
      />
      <span
        id="aria-context"
      >
         7 results available. Use Up and Down to choose options, press Enter to select the currently focused option, press Escape to exit the menu, press Tab to select the option and exit the menu.
      </span>
    </span>
    <div
      class="CalculationOptions__control CalculationOptions__control--is-focused CalculationOptions__control--menu-is-open css-1s59geg-Control"
    >
      <div
        class="CalculationOptions__value-container CalculationOptions__value-container--has-value css-1mxrbau-ValueContainer"
      >
        <div
          class="CalculationOptions__single-value css-1brck82-singleValue"
        >
          Count
        </div>
        <input
          aria-autocomplete="list"
          class="css-wmatm6-dummyInput-DummyInput"
          id="react-select-4-input"
          readonly=""
          tabindex="0"
          value=""
        />
      </div>
      <div
        class="CalculationOptions__indicators css-1hb7zxy-IndicatorsContainer"
      >
        <div
          aria-hidden="true"
          class="CalculationOptions__indicator CalculationOptions__clear-indicator css-13eygzs-indicatorContainer"
        >
          <svg
            aria-hidden="true"
            class="css-tj5bde-Svg"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <span
          class="CalculationOptions__indicator-separator css-43ykx9-indicatorSeparator"
        />
        <div
          aria-hidden="true"
          class="CalculationOptions__indicator CalculationOptions__dropdown-indicator css-1f9iddu-indicatorContainer"
        >
          <i
            class="CompassIcon icon-chevron-up ChevronUpIcon"
          />
        </div>
      </div>
    </div>
    <div
      class="CalculationOptions__menu css-1rsmi4x-menu"
    >
      <div
        class="CalculationOptions__menu-list css-g29tl0-MenuList"
      >
        <div
          class="KanbanCalculationOptions_CustomOption active"
        >
          <span>
            Count
             
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Not Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Percent Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Percent Not Empty
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Value
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
        </div>
        <div
          class="KanbanCalculationOptions_CustomOption "
        >
          <span>
            Count Unique Values
             
            <i
              class="CompassIcon icon-chevron-right ChevronRightIcon"
            />
          </span>
          <div
            class="dropdown-submenu"
          >
            <div
              class="drops "
            >
              <span>
                Status
              </span>
            </div>
            <div
              class="drops active"
            >
              <span>
                Property 1
              </span>
            </div>
            <div
              class="drops "
            >
              <span>
                Property 2
              </span>
            </div>
            <div
              class="drops "
            >
              <span>
                Property 3
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
    <input
      name="calculation_options"
      type="hidden"
      value="count"
    />
  </div>
</div>
`;
